src/lib/uri/uri.c

```diff
@@ -371,3 +372,92 @@ uri_set_create(struct uri_set *uri_set, const char *str)
 	uri_set->uri_count = 0;
 	return -1;
 }
+
+static unsigned char unreserved_symbols[256] = {0};
+
+/*
+ * Builds an array with unreserved characters.
+ */
+static void
+build_unreserved_symbols_table(const char *str)
+{
+	for (int i = 0; i < 256; i++)
+		unreserved_symbols[i] = 0;
+
+	for (; *str; str++) {
+		unsigned char ch = (unsigned char)*str;
+		unreserved_symbols[ch] = 1;
+	}
+}
+
+/*
+ * String percent-encoding.
+ */
+char *
+uri_escape(const char *buf, size_t len, const char *unreserved, bool encode_plus)
+{
+	build_unreserved_symbols_table(unreserved);
+	/* The worst case is when all characters are encoded. */
+	char *encoded = xcalloc(len * 3 + 1, sizeof(char));
+	int pos = 0;
+	const char *hex = "0123456789ABCDEF";
+	while (len--) {
+		unsigned char ch = (unsigned char)*buf;
+		if ((ch == ' ') && encode_plus) {
+			encoded[pos++] = '+';
+		} else if (!unreserved_symbols[ch]) {
+			encoded[pos++] = '%';
+			encoded[pos++] = hex[ch >> 4];
+			encoded[pos++] = hex[ch & 15];
+		} else {
+			encoded[pos++] = *buf;
+		}
+		buf++;
+	}
+	encoded[pos++] = '\0';
+	return encoded;
+};
+
+/*
+ * Converts a hex character to its integer value.
+ */
+static char
+hex2ch(char ch)
+{
+	return isdigit(ch) ? ch - '0' : tolower(ch) - 'a' + 10;
```

Comment:
`man isdigit` and `man tolower` say:

> ## NOTES
>
> The standards require that the argument `c` for these functions is either `EOF` or a value that is representable in the type `unsigned char`.  If the argument `c` is of type `char`, it must be cast to unsigned char, as in the following example:
>
> ```
> char c;
> ...
> res = toupper((unsigned char) c);
> ```
>
> This is necessary because `char` may be the equivalent of `signed char`, in which case a byte where the top bit is set would be sign extended when converting to `int`, yielding a value that is outside the range of `unsigned char`.

Reply:
The only information I have is what the man page says. I guess that `char` -> `unsigned char` implicit cast at passing the argument to `hex2ch()` should do the trick.